README.md
# panz-c8

panz-c8 is an emulator for the CHIP-8 system written in C. At the moment the emulator doesn't support sound.

## Build

panz-c8 requires [SDL 2.0](http://www.libsdl.org). In the repository there is a makefile for build the emulator on OSX, but it should be easy to build the emulator using any C99 compiler.


CHANGELOG.md
# panz-c8 1.0.2

  - New repository structure

# panz-c8 1.0.1

  - Updated LICENSE
  - New makefile
  - Some little code cleaning and updates

# panz-c8 1.0

  - First public release of panz-c8

c8-cpu.c
#include "cpu.h"
#include <string.h>
#include <stdio.h>

void initCPU(c8_cpu *cpu) {
    unsigned char chip8_fontset[80] = {
        0xF0, 0x90, 0x90,
        0x90, 0xF0, // 0
        0x20, 0x60, 0x20,
        0x20, 0x70, // 1
        0xF0, 0x10, 0xF0,
        0x80, 0xF0, // 2
        0xF0, 0x10, 0xF0,
        0x10, 0xF0, // 3
        0x90, 0x90, 0xF0,
        0x10, 0x10, // 4
        0xF0, 0x80, 0xF0,
        0x10, 0xF0, // 5
        0xF0, 0x80, 0xF0,
        0x90, 0xF0, // 6
        0xF0, 0x10, 0x20,
        0x40, 0x40, // 7
        0xF0, 0x90, 0xF0,
        0x90, 0xF0, // 8
        0xF0, 0x90, 0xF0,
        0x10, 0xF0, // 9
        0xF0, 0x90, 0xF0,
        0x90, 0x90, // A
        0xE0, 0x90, 0xE0,
        0x90, 0xE0, // B
        0xF0, 0x80, 0x80,
        0x80, 0xF0, // C
        0xE0, 0x90, 0x90,
        0x90, 0xE0, // D
        0xF0, 0x80, 0xF0,
        0x80, 0xF0, // E
        0xF0, 0x80, 0xF0,
        0x80, 0x80 // F
    };

    memset(cpu->memory, 0, MEM_SIZE);
    memcpy(cpu->memory, chip8_fontset, 80);
    memset(cpu->registers, 0, 16);
    cpu->mem_register = 0;
    cpu->prog_counter = 0x200;
    cpu->stack = (WORD *)(cpu->memory + 0xEFF);
    memset(cpu->keymap, 0, 16);
    memset(cpu->video_memory, 0, 64 * 32);

    cpu->timer = 0;
    cpu->soundTimer = 0;
    cpu->randomByte = 0;
}

void setRandomByte(c8_cpu *cpu, BYTE value) {
    cpu->randomByte = value;
}

WORD fetch(c8_cpu *cpu) {
    WORD opcode = 0;
    opcode = cpu->memory[cpu->prog_counter];
    opcode <<= 8;
    opcode |= cpu->memory[cpu->prog_counter + 1];
    cpu->prog_counter += 2;

    return opcode;
}

void decode(WORD opcode, c8_cpu *cpu) {
    WORD first = FIRST_HEX(opcode);
    switch (first) {
    case 0x0:
        execute0Opc(opcode, cpu);
        break;
    case 0x1:
        execute1NNN(opcode, cpu);
        break;
    case 0x2:
        execute2NNN(opcode, cpu);
        break;
    case 0x3:
        execute3XNN(opcode, cpu);
        break;
    case 0x4:
        execute4XNN(opcode, cpu);
        break;
    case 0x5:
        execute5XY0(opcode, cpu);
        break;
    case 0x6:
        execute6XNN(opcode, cpu);
        break;
    case 0x7:
        execute7XNN(opcode, cpu);
        break;
    case 0x8:
        execute8Opc(opcode, cpu);
        break;
    case 0x9:
        execute9XY0(opcode, cpu);
        break;
    case 0xA:
        executeANNN(opcode, cpu);
        break;
    case 0xB:
        executeBNNN(opcode, cpu);
        break;
    case 0xC:
        executeCXNN(opcode, cpu);
        break;
    case 0xD:
        executeDXYN(opcode, cpu);
        break;
    case 0xE:
        executeEOpc(opcode, cpu);
        break;
    case 0xF:
        executeFOpc(opcode, cpu);
        break;
    default:
        break;
    };
}

void execute0Opc(WORD opcode, c8_cpu *cpu) {
    WORD second = SEC_HEX(opcode);

    if (second == 0) {
        WORD four = FOUR_HEX(opcode);
        if (four == 0)
            execute00E0(opcode, cpu);
        else
            execute00EE(opcode, cpu);
    } else
        execute0NNN(opcode, cpu);
}

void execute8Opc(WORD opcode, c8_cpu *cpu) {
    WORD four = FOUR_HEX(opcode);
    switch (four) {
    case 0x0:
        execute8XY0(opcode, cpu);
        break;
    case 0x1:
        execute8XY1(opcode, cpu);
        break;
    case 0x2:
        execute8XY2(opcode, cpu);
        break;
    case 0x3:
        execute8XY3(opcode, cpu);
        break;
    case 0x4:
        execute8XY4(opcode, cpu);
        break;
    case 0x5:
        execute8XY5(opcode, cpu);
        break;
    case 0x6:
        execute8XY6(opcode, cpu);
        break;
    case 0x7:
        execute8XY7(opcode, cpu);
        break;
    case 0xE:
        execute8XYE(opcode, cpu);
        break;
    default:
        break;
    };
}

void executeEOpc(WORD opcode, c8_cpu *cpu) {
    WORD third = THIRD_HEX(opcode);
    switch (third) {
    case 0x9:
        executeEX9E(opcode, cpu);
        break;
    case 0xA:
        executeEXA1(opcode, cpu);
        break;
    default:
        break;
    };
}

void executeFOpc(WORD opcode, c8_cpu *cpu) {
    WORD value = opcode & 0xFF;
    switch (value) {
    case 0x07:
        executeFX07(opcode, cpu);
        break;
    case 0x0A:
        executeFX0A(opcode, cpu);
        break;
    case 0x15:
        executeFX15(opcode, cpu);
        break;
    case 0x18:
        executeFX18(opcode, cpu);
        break;
    case 0x1E:
        executeFX1E(opcode, cpu);
        break;
    case 0x29:
        executeFX29(opcode, cpu);
        break;
    case 0x33:
        executeFX33(opcode, cpu);
        break;
    case 0x55:
        executeFX55(opcode, cpu);
        break;
    case 0x65:
        executeFX65(opcode, cpu);
        break;
    default:
        break;
    }
}

void updateTimer(c8_cpu *cpu) {
    if (cpu->timer > 0)
        cpu->timer--;
    if (cpu->soundTimer > 0)
        cpu->soundTimer--;
}


opcode.c
#include "cpu.h"
#include <string.h>

void execute0NNN(WORD opcode, c8_cpu *cpu) {
}

void execute00E0(WORD opcode, c8_cpu *cpu) {
    memset(cpu->video_memory, 0, 32 * 64);
}

void execute00EE(WORD opcode, c8_cpu *cpu) {
    WORD val;
    POP(cpu->stack, val);
    cpu->prog_counter = val;
}

void execute1NNN(WORD opcode, c8_cpu *cpu) {
    WORD addr = opcode & 0xFFF;
    cpu->prog_counter = addr;
}

void execute2NNN(WORD opcode, c8_cpu *cpu) {
    WORD addr = opcode & 0xFFF;
    PUSH(cpu->stack, cpu->prog_counter);
    cpu->prog_counter = addr;
}

void execute3XNN(WORD opcode, c8_cpu *cpu) {
    WORD register_num = SEC_HEX(opcode);
    WORD value = opcode & 0xFF;
    if (cpu->registers[register_num] == value)
        cpu->prog_counter += 2;
}

void execute4XNN(WORD opcode, c8_cpu *cpu) {
    WORD register_num = SEC_HEX(opcode);
    WORD value = opcode & 0xFF;
    if (cpu->registers[register_num] != value)
        cpu->prog_counter += 2;
}

void execute5XY0(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    if (cpu->registers[fir] == cpu->registers[sec])
        cpu->prog_counter += 2;
}

void execute6XNN(WORD opcode, c8_cpu *cpu) {
    WORD register_num = SEC_HEX(opcode);
    WORD value = opcode & 0xFF;
    cpu->registers[register_num] = value;
}

void execute7XNN(WORD opcode, c8_cpu *cpu) {
    WORD register_num = SEC_HEX(opcode);
    WORD value = opcode & 0xFF;
    cpu->registers[register_num] += value;
}

void execute8XY0(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[fir] = cpu->registers[sec];
}

void execute8XY1(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[fir] |= cpu->registers[sec];
}

void execute8XY2(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[fir] &= cpu->registers[sec];
}

void execute8XY3(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[fir] ^= cpu->registers[sec];
}

void execute8XY4(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[15] = 0;

    WORD value = cpu->registers[fir] + cpu->registers[sec];
    if (value > 255)
        cpu->registers[15] = 1;

    cpu->registers[fir] += cpu->registers[sec];
}

void execute8XY5(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[15] = 1;

    if (cpu->registers[fir] < cpu->registers[sec])
        cpu->registers[15] = 0;

    cpu->registers[fir] -= cpu->registers[sec];
}

void execute8XY6(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    cpu->registers[15] = cpu->registers[fir] & 0x1;
    cpu->registers[fir] >>= 1;
}

void execute8XY7(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    cpu->registers[15] = 1;

    if (cpu->registers[sec] < cpu->registers[fir])
        cpu->registers[15] = 0;

    cpu->registers[fir] = cpu->registers[sec] - cpu->registers[fir];
}

void execute8XYE(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    cpu->registers[15] = cpu->registers[fir] >> 7;
    cpu->registers[fir] <<= 1;
}

void execute9XY0(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);
    if (cpu->registers[fir] != cpu->registers[sec])
        cpu->prog_counter += 2;
}

void executeANNN(WORD opcode, c8_cpu *cpu) {
    WORD value = opcode & 0xFFF;
    cpu->mem_register = value;
}

void executeBNNN(WORD opcode, c8_cpu *cpu) {
    WORD value = opcode & 0xFFF;
    cpu->prog_counter = value + cpu->registers[0];
}

void executeCXNN(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD mask = opcode & 0xFF;

    cpu->registers[fir] = cpu->randomByte & mask; // TODO: stub, should be
                                                  // random
}

void executeDXYN(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    WORD sec = THIRD_HEX(opcode);

    BYTE coordx = cpu->registers[fir];
    BYTE coordy = cpu->registers[sec];
    BYTE height = opcode & 0xF;

    cpu->registers[15] = 0;

    for (BYTE yline = 0; yline < height; yline++) {
        BYTE data = cpu->memory[cpu->mem_register + yline];

        BYTE xpixel = 0;
        BYTE xpixelinv = 7;
        for (xpixel = 0; xpixel < 8; xpixel++, xpixelinv--) {
            BYTE work = data >> xpixelinv;
            work &= 0x1;
            BYTE x = xpixel + coordx;
            BYTE y = coordy + yline;
            BYTE last = cpu->video_memory[y][x];
            cpu->video_memory[y][x] ^= work;
            if ((last == 1) && (cpu->video_memory[y][x] == 0))
                cpu->registers[15] = 1;
        }
    }
}

void executeEX9E(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    BYTE val = cpu->registers[fir];
    if (cpu->keymap[val] == 1)
        cpu->prog_counter += 2;
}

void executeEXA1(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    BYTE val = cpu->registers[fir];
    if (cpu->keymap[val] == 0)
        cpu->prog_counter += 2;
}

void executeFX07(WORD opcode, c8_cpu *cpu) {
    WORD val = SEC_HEX(opcode);
    cpu->registers[val] = cpu->timer;
}

void executeFX0A(WORD opcode, c8_cpu *cpu) {
    WORD val = SEC_HEX(opcode);
    BYTE i;

    for (i = 0; i < 16; i++) {
        if (cpu->keymap[i] > 0)
            break;
    }

    if (i == 16)
        cpu->prog_counter -= 2;
    else
        cpu->registers[val] = i;
}

void executeFX15(WORD opcode, c8_cpu *cpu) {
    WORD val = SEC_HEX(opcode);
    cpu->timer = cpu->registers[val];
}

void executeFX18(WORD opcode, c8_cpu *cpu) {
    WORD val = SEC_HEX(opcode);
    cpu->soundTimer = cpu->registers[val];
}

void executeFX1E(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    cpu->mem_register += cpu->registers[fir];
}

void executeFX29(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    cpu->mem_register = cpu->registers[fir] * 5;
}

void executeFX33(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);

    BYTE val = cpu->registers[fir];
    cpu->memory[cpu->mem_register] = val / 100;
    cpu->memory[cpu->mem_register + 1] = (val / 10) % 10;
    cpu->memory[cpu->mem_register + 2] = val % 10;
}

void executeFX55(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    BYTE i;
    for (i = 0; i <= fir; i++)
        cpu->memory[cpu->mem_register + i] = cpu->registers[i];
    cpu->mem_register += fir;
    cpu->mem_register += 1;
}

void executeFX65(WORD opcode, c8_cpu *cpu) {
    WORD fir = SEC_HEX(opcode);
    BYTE i;
    for (i = 0; i <= fir; i++)
        cpu->registers[i] = cpu->memory[cpu->mem_register + i];
    cpu->mem_register += fir;
    cpu->mem_register += 1;
}


src/cpu.h
#ifndef CPU_H
#define CPU_H

#include <stdint.h>

typedef uint8_t BYTE;
typedef uint16_t WORD;

#define MEM_SIZE 0xFFF

#define FIRST_HEX(a) ((a & 0xF000) >> 12)
#define SEC_HEX(a) ((a & 0xF00) >> 8)
#define THIRD_HEX(a) ((a & 0xF0) >> 4)
#define FOUR_HEX(a) ((a & 0xF))

#define PUSH(base, l) *(base--) = l;
#define POP(base, l) l = (*++base);

// 76543210
#define GET_BIT(byte, numBit) (byte & (0x1 << numBit)) >> numBit

typedef struct c8_cpu {
    BYTE memory[MEM_SIZE];
    BYTE registers[16];
    WORD mem_register;
    WORD prog_counter;

    WORD timer;
    WORD soundTimer;

    WORD *stack;

    /*All these variables should me mapped on memory*/

    BYTE keymap[16];
    BYTE video_memory[32][64];
    BYTE randomByte;
} c8_cpu;

void initCPU(c8_cpu *cpu);
void updateTimer(c8_cpu *cpu);
void setRandomByte(c8_cpu *cpu, BYTE value);

WORD fetch(c8_cpu *cpu);
void decode(WORD opcode, c8_cpu *cpu);

void execute8Opc(WORD opcode, c8_cpu *cpu);
void execute0Opc(WORD opcode, c8_cpu *cpu);
void executeEOpc(WORD opcode, c8_cpu *cpu);
void executeFOpc(WORD opcode, c8_cpu *cpu);

void execute0NNN(WORD opcode, c8_cpu *cpu);
void execute00E0(WORD opcode, c8_cpu *cpu);
void execute00EE(WORD opcode, c8_cpu *cpu);
void execute1NNN(WORD opcode, c8_cpu *cpu);
void execute2NNN(WORD opcode, c8_cpu *cpu);
void execute3XNN(WORD opcode, c8_cpu *cpu);
void execute4XNN(WORD opcode, c8_cpu *cpu);
void execute5XY0(WORD opcode, c8_cpu *cpu);
void execute6XNN(WORD opcode, c8_cpu *cpu);
void execute7XNN(WORD opcode, c8_cpu *cpu);
void execute8XY0(WORD opcode, c8_cpu *cpu);
void execute8XY1(WORD opcode, c8_cpu *cpu);
void execute8XY2(WORD opcode, c8_cpu *cpu);
void execute8XY3(WORD opcode, c8_cpu *cpu);
void execute8XY4(WORD opcode, c8_cpu *cpu);
void execute8XY5(WORD opcode, c8_cpu *cpu);
void execute8XY6(WORD opcode, c8_cpu *cpu);
void execute8XY7(WORD opcode, c8_cpu *cpu);
void execute8XYE(WORD opcode, c8_cpu *cpu);
void execute9XY0(WORD opcode, c8_cpu *cpu);
void executeANNN(WORD opcode, c8_cpu *cpu);
void executeBNNN(WORD opcode, c8_cpu *cpu);
void executeCXNN(WORD opcode, c8_cpu *cpu);
void executeDXYN(WORD opcode, c8_cpu *cpu);
void executeEX9E(WORD opcode, c8_cpu *cpu);
void executeEXA1(WORD opcode, c8_cpu *cpu);
void executeFX07(WORD opcode, c8_cpu *cpu);
void executeFX0A(WORD opcode, c8_cpu *cpu);
void executeFX15(WORD opcode, c8_cpu *cpu);
void executeFX18(WORD opcode, c8_cpu *cpu);
void executeFX1E(WORD opcode, c8_cpu *cpu);
void executeFX29(WORD opcode, c8_cpu *cpu);
void executeFX33(WORD opcode, c8_cpu *cpu);
void executeFX55(WORD opcode, c8_cpu *cpu);
void executeFX65(WORD opcode, c8_cpu *cpu);

#endif


src/main.c
#include <SDL2/SDL.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include "cpu.h"

#define RES_W 64
#define RES_H 32

void loadROM(c8_cpu *cpu, char *rom) {
    FILE *f = fopen(rom, "rb");
    if (!f) {
        fprintf(stderr, "Can't load ROM\n");
        exit(2);
    }
    fread(cpu->memory + 0x200, 1, MEM_SIZE - 0x200, f);
    fclose(f);
}

void renderVideoMemory(c8_cpu *cpu, SDL_Renderer *rend, int SCALE) {
    int i = 0, j = 0;
    SDL_SetRenderDrawColor(rend, 0, 0, 0, 0);
    SDL_RenderClear(rend);
    SDL_Rect rectangle;
    rectangle.w = SCALE;
    rectangle.h = SCALE;
    SDL_SetRenderDrawColor(rend, 255, 255, 255, 0);
    for (i = 0; i < 32; i += 1)
        for (j = 0; j < 64; j += 1)
            if (cpu->video_memory[i][j] != 0) {
                rectangle.x = j * SCALE;
                rectangle.y = i * SCALE;
                SDL_RenderFillRect(rend, &rectangle);
            }
    SDL_RenderPresent(rend);
}

void getInput(c8_cpu *cpu) {
    SDL_Event event;
    int value, key;
    while (SDL_PollEvent(&event)) {
        if (event.type == SDL_QUIT)
            exit(EXIT_SUCCESS);
        if (event.type == SDL_KEYDOWN || event.type == SDL_KEYUP) {
            switch (event.key.keysym.sym) {
            case SDLK_x:
                key = 0;
                break;
            case SDLK_1:
                key = 1;
                break;
            case SDLK_2:
                key = 2;
                break;
            case SDLK_3:
                key = 3;
                break;
            case SDLK_q:
                key = 4;
                break;
            case SDLK_w:
                key = 5;
                break;
            case SDLK_e:
                key = 6;
                break;
            case SDLK_a:
                key = 7;
                break;
            case SDLK_s:
                key = 8;
                break;
            case SDLK_d:
                key = 9;
                break;
            case SDLK_z:
                key = 10;
                break;
            case SDLK_c:
                key = 11;
                break;
            case SDLK_4:
                key = 12;
                break;
            case SDLK_r:
                key = 13;
                break;
            case SDLK_f:
                key = 14;
                break;
            case SDLK_v:
                key = 15;
                break;
            default:
                key = -1;
                break;
            }
            if (event.type == SDL_KEYDOWN)
                value = 1;
            else
                value = 0;

            if (key != -1)
                cpu->keymap[key] = value;
        }
    }
}

void handleArguments(char *configFile, unsigned int *configOption,
                     c8_cpu *cpu) {
    FILE *config = fopen(configFile, "rt");
    if (!config) {
        fprintf(stderr, "Can't load config.cfg file\n");
        exit(EXIT_FAILURE);
    }

    char buffer[128];

    while (fscanf(config, "%s128\n", buffer) == 1) {
        char *variableName = strtok(buffer, "*");
        if (strcmp(variableName, "SCALE") == 0) {
            configOption[0] = atoi(strtok(NULL, "*"));
        } else if (strcmp(variableName, "OPCODES") == 0) {
            configOption[1] = atoi(strtok(NULL, "*"));
        }
    }
}

void initConfig(unsigned int *config) {
    config[0] = 1;
    config[1] = 400;
}

int main(int argc, char **argv) {
    c8_cpu c8;
    SDL_Renderer *renderer;
    SDL_Window *window;
    /*configOption:
     * 0 is the resolution scale
     * 1 is the number of opcode per second
     * 2 is a flag for the sound
     */
    unsigned int configOption[3];
    unsigned int *SCALE = configOption;
    unsigned int *opcodePerSec = configOption + 1;

    initConfig(configOption);
    initCPU(&c8);
    SDL_Init(SDL_INIT_EVERYTHING);

    handleArguments("config.cfg", configOption, &c8);
    loadROM(&c8, argv[1]);

    window = SDL_CreateWindow("panz-c8", SDL_WINDOWPOS_UNDEFINED,
                              SDL_WINDOWPOS_UNDEFINED, RES_W * *SCALE,
                              RES_H * *SCALE, 0);
    renderer = SDL_CreateRenderer(window, -1, 0);
    while (1) {
        unsigned int timeStartFrame = SDL_GetTicks();
        for (int i = 0; i < *opcodePerSec / 60; i++) {
            getInput(&c8);
            WORD opcode = fetch(&c8);
            decode(opcode, &c8);
            setRandomByte(&c8, SDL_GetTicks() & 0xFF);
        }
        updateTimer(&c8);
        renderVideoMemory(&c8, renderer, *SCALE);
        float deltaT =
            (float)1000 / 60 - (float)(SDL_GetTicks() - timeStartFrame);
        if (deltaT > 0)
            SDL_Delay((unsigned int)deltaT);
    }
}
